### PR TITLE
Clean-up BasicRest template so that it's easier to create new ones from CreateProject.

### PR DESCRIPTION
**Pull request by metacosm**
This PR started out as a simplification of how Quarkus handled source types (e.g. java vs kotlin) by moving source type specific information to the `SourceType` enumeration. It evolved into proposing a strategy (using `ServiceLoader`) and a conventional layout for project templates.

Currently, templates for sources and resources can be specified for the chosen project template (e.g. `basic-rest`) by moving the templates from `templates` to `templates/{name of project template}/{target language}` (so existing templates are moved to `templates/basic-rest/java` and `templates/basic-rest/kotlin`). This should opens up the possibility to create different types of projects using project templates though the real benefit would come from the fact that templates are loaded dynamically using `ServiceLoader` and should therefore be moved out of the `common` module and packaged into their own artifact(s). 

I still have ideas on improvements but would rather do them in a separate PR:

- Move template code into their own module/project.
- Move more code to `CreateProject` (already did some regarding class and package name) so that common functionality is handle at the command level instead of the template one (share more code).
- Introduce an `AbstractQuarkusTemplate` class that template implementors could reuse if it makes sense (depending on the amount of code that can be moved to `CreateProject`)

Ideas / comments?

### CODE REVIEW

**Comment by geoand** [comment]
I for one really like this cleanup!

**Comment by gsmet** [comment]
@geoand I agree.

@cescoffier could you take a look? It's more your area of expertise.

**Comment by gsmet** [comment]
@metacosm btw, could you rebase? I had to fix some issues in the default template.

**Comment by metacosm** [comment]
> @metacosm btw, could you rebase? I had to fix some issues in the default template.

Rebased, I was expecting conflicts from your PR… :)

**Comment by gsmet** [comment]
@metacosm just to let you know that @cescoffier is afk right now. We'll get back to you soon.

**Comment by metacosm** [comment]
@gsmet no problems, I know @cescoffier is busy! :)

**Comment by gsmet** [comment]
@metacosm sorry, broke it again, could you rebase? Last time, hopefully.

**Comment by metacosm** [comment]
> @metacosm sorry, broke it again, could you rebase? Last time, hopefully.

@gsmet do you have a link to the commit that broke this? Rebasing locally doesn't show any conflicts so I want to manually double-check… Also, any ETA on a review?

**Comment by metacosm** [comment]
@gsmet rebased

**Comment by metacosm** [comment]
Rebased to pick up changes to fix #1565

**Comment by metacosm** [comment]
ping @cescoffier

**Comment by gsmet** [comment]
@metacosm sorry, it needs a rebase again :/

Clément is back but is busy preparing tomorrow's release. He'll get back to you shortly.

**Comment by metacosm** [comment]
We need a bot to track this rebase needs… :)

File: devtools/common/src/main/java/io/quarkus/templates/TemplateRegistry.java
Comment (cescoffier) [inline devtools/common/src/main/java/io/quarkus/templates/TemplateRegistry.java]: Shouldn't it throw a runtime exception if the template is null?

File: devtools/common/src/main/java/io/quarkus/templates/TemplateRegistry.java
Comment (cescoffier) [inline devtools/common/src/main/java/io/quarkus/templates/TemplateRegistry.java]: what about NoSuchElementException?

File: devtools/common/src/main/java/io/quarkus/templates/rest/BasicRest.java
Comment (cescoffier) [inline devtools/common/src/main/java/io/quarkus/templates/rest/BasicRest.java]: Wondering here... We should forbid the default package.

File: devtools/common/src/main/java/io/quarkus/templates/rest/BasicRest.java
Comment (cescoffier) [inline devtools/common/src/main/java/io/quarkus/templates/rest/BasicRest.java]: Any reason why some variables are final and some are not? 

I don't think it adds anything there.

**Review by cescoffier — CHANGES_REQUESTED** [review]
Looks nice! I've made a few comments.

File: devtools/common/src/main/java/io/quarkus/templates/TemplateRegistry.java
Comment (metacosm) [reply devtools/common/src/main/java/io/quarkus/templates/TemplateRegistry.java]: Probably, yes.

File: devtools/common/src/main/java/io/quarkus/templates/rest/BasicRest.java
Comment (metacosm) [reply devtools/common/src/main/java/io/quarkus/templates/rest/BasicRest.java]: Force of habit. I make everything final when I can but I agree that it doesn't add anything (apart from extra words) :)

**Comment by metacosm** [comment]
@cescoffier updated.

**Comment by gsmet** [comment]
@cescoffier @metacosm btw, once things are settled and before merging, please try to squash things a bit to have atomic commits.

**Comment by metacosm** [comment]
@gsmet don't you squash and merge when merging?

**Comment by gsmet** [comment]
@metacosm so, I don't personally like it, but we use merge commits. And I don't think GH has an option to squash and merge when using merge commit.

**Comment by metacosm** [comment]
@gsmet ok, I'll squash if I get the greenlight for a merge

**Review by cescoffier — APPROVED** [review]

**Pull request merged**
